java/flight/src/main/java/org/apache/arrow/flight/FlightStream.java

```diff
@@ -56,10 +67,12 @@
   private volatile int pending = 1;
   private boolean completed = false;
   private volatile VectorSchemaRoot fulfilledRoot;
+  private DictionaryProvider.MapDictionaryProvider dictionaries;
   private volatile VectorLoader loader;
   private volatile Throwable ex;
   private volatile FlightDescriptor descriptor;
   private volatile Schema schema;
+  private volatile byte[] applicationMetadata = null;
```

Comment:
Let's construct as a listener pattern instead. Assuming people only want the last one seems a bit specific to what must be your use case.

Reply:
Okay. I initially thought it would make correlating data with the associated metadata hard (and in the protocol, they're sent together - the metadata is not out-of-band for the cases that FlightStream represents), but the Java APIs are not really designed around manipulating individual RecordBatches anyways, so a listener would make sense.

Reply:
@jacques-n on second thought, I don't think a listener is appropriate - it complicates usage quite a bit, especially if you want to correlate a particular metadatum with a particular RecordBatch. The current code does guarantee that `applicationMetadata` is not updated except by a call to `next`, which is the same invariant as for the `VectorSchemaRoot` contained in `FlightStream`. So if you do want previous metadata messages, you can retain them yourself (as you would with a listener anyways).